olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/util/SubprocessConnection.java

```diff
@@ -192,112 +292,230 @@ public void setTimeout(int time, TimeUnit unit) {
+     */
+    public void shutdown(boolean graceful) {
+        try {
+            processLock.lock();
+            if (process != null && process.isAlive()) {
+                //
+                // Announce that we're about to shut down.
+                listeners.forEach(l -> l.subprocessPreShutdown(this));
+
+                try {
+                    if (graceful) {
+                        // First, tell it to shut down, then wait a short
+                        // period before making sure it is down.
+                        PrintWriter stdin = new PrintWriter(process.getOutputStream());
+                        stdin.println(SHUTDOWN);
+                        stdin.flush();
                         if (!process.waitFor(5, TimeUnit.SECONDS)) {
                             process.destroyForcibly();
                         }
-                    } catch (InterruptedException e) {
-                        logger.log(Level.WARNING, "Shutdown interrupted", e);
+                    } else {
                         process.destroyForcibly();
                     }
-                    //
-                    // Announce that shutdown has happened.
-                    listeners.forEach(l -> l.subprocessPostShutdown(this));
+                } catch (InterruptedException e) {
+                    logger.log(Level.WARNING, "Shutdown interrupted", e);
+                    process.destroyForcibly();
+                } finally {
+                    process = null;
                 }
+                //
+                // Announce that shutdown has happened.
+                listeners.forEach(l -> l.subprocessPostShutdown(this));
+            } else {
+                //
+                // Announce that shutdown has happened.
+                listeners.forEach(l -> l.subprocessPostShutdown(this));
             }
+            if (idlerTimer != null) {
+                idlerTimer.cancel();
+                idlerTimer = null;
+            }
+        } finally {
+            processLock.unlock();
+        }
+
+    }
+
+    /**
+     * Destroys a subprocess while handling appropriate locking
+     */
+    protected class SubprocessReaper extends TimerTask {
+        protected Process proc;
+
+        public SubprocessReaper(Process proc) {
+            this.proc = proc;
         }
-        if (idlerTimer != null) {
-            idlerTimer.cancel();
-            idlerTimer = null;
+
+        @Override
+        public void run() {
+            //
+            // Announce that we're about to shut down.
+            listeners.forEach(l -> l.subprocessPreShutdown(SubprocessConnection.this));
+
+            logger.fine("Killing subprocess");
+            try {
+                proc.destroyForcibly();
+                //
+                // Get in line to try to be the next one to hold the lock. This way nobody can
+                // reclaim it until the process is fully exited.
+                processLock.lock();
```

Comment:
Shouldn't it lock before `destroyForcibly()` is called? As otherwise something else could have the lock and not expect the process to go away.

Reply:
The reason for this is that this thread needs to be able to destroy the process while it is running.  The Reaper handles the situation where the subprocess has gone away for too long and we're assuming it is dead.  The reaper runs while the process is running but (obviously) in a separate thread so it would deadlock if it tried to grab the `processLock` first.  Forcibly destroying the process will cause an exception to be thrown up in the run method and the process lock will be released at that point. Since `destroyForcibly()` returns asynchronously, the reaper relies on the fact that `run` is synchronized to make sure that its locking should happen before any other threads could get into run and obtain the process lock.  If it fails to do that, then the next call to `run` could throw an IOException depending on the timing.  I'm adding some more comments explaining this.